sdk/iot/Azure.Iot.Hub.Service/tests/QueryTests.cs

```diff
@@ -0,0 +1,146 @@
+                        {
+                            DeviceId = testDeviceId
+                        })
+                    .ConfigureAwait(false))
+                    .Value;
+
+                // Adding a delay to account for query cache sync.
+                // TODO: Figure out a better way to ensure the test always succeeds.
+                await Task.Delay(TimeSpan.FromSeconds(5)).ConfigureAwait(false);
+
+                // Query for device twins with a specific tag.
+                AsyncPageable<TwinData> queryResponse = client.Query.QueryAsync($"SELECT * FROM devices WHERE deviceId = '{testDeviceId}'");
+
+                int count = 0;
+                await foreach (TwinData item in queryResponse)
+                {
+                    count++;
+                }
+
+                count.Should().Be(1);
+
+                // Delete the device
+                // Deleting the device happens in the finally block as cleanup.
+            }
+            finally
+            {
+                await CleanupAsync(client, device).ConfigureAwait(false);
+            }
+        }
+
+        /// <summary>
+        /// Test querying all module twins in the IoTHub.
+        /// </summary>
+        [Test]
+        public async Task QueryClient_GetAllModulesTwinsWithTag()
+        {
+            string testDeviceId = $"QueryDevice{GetRandom()}";
+            string testModuleId = $"QueryModule{GetRandom()}";
+
+            DeviceIdentity device = null;
+            IotHubServiceClient client = GetClient();
+
+            try
+            {
+                // Create a device to house the module
+                device = (await client.Devices
+                    .CreateOrUpdateIdentityAsync(
+                        new DeviceIdentity
+                        {
+                            DeviceId = testDeviceId
+                        })
+                    .ConfigureAwait(false))
+                    .Value;
+
+                // Create a module on the device
+                Response<ModuleIdentity> createResponse = await client.Modules
+                    .CreateOrUpdateIdentityAsync(
+                        new ModuleIdentity
+                        {
+                            DeviceId = testDeviceId,
+                            ModuleId = testModuleId
+                        })
+                    .ConfigureAwait(false);
+
+                // Adding a delay to account for query cache sync.
+                // TODO: Figure out a better way to ensure the test always succeeds.
+                await Task.Delay(TimeSpan.FromSeconds(5)).ConfigureAwait(false);
+
+                // Query for module twins with a specific tag.
+                AsyncPageable<TwinData> queryResponse = client.Query.QueryAsync($"SELECT * FROM devices.modules WHERE moduleId = '{testModuleId}'");
+
+                int count = 0;
+                await foreach (TwinData item in queryResponse)
+                {
+                    count++;
```

Comment:
You could add a check here that the returned item has the expected moduleId just to make sure the query is working as expected

Reply:
Good point